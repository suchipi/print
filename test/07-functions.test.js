const { print } = require("./helpers");

describe("Functions", () => {
	it("prints source code", () => {
		function func(i) {
			return i + 2;
		}
		expect(print(func)).toMatchSnapshot();
		expect(print(() => "A")).toMatchSnapshot();
		const obj = {
			func: function fn(value) {
				return value;
			},
			arrow: (a, b) => a + b,
		};
		expect(print(obj)).toMatchSnapshot();
	});

	it("identifies generator functions", () => {
		function* foo() {
			yield 20;
		}
		expect(print(foo)).toMatchSnapshot();
		expect(print({ method: foo })).toMatchSnapshot();
	});

	it("identifies asynchronous functions", () => {
		async function foo() {
			return 20;
		}
		expect(print(foo)).toMatchSnapshot();
		expect(print({ method: foo })).toMatchSnapshot();
	});

	it("identifies asynchronous generators", () => {
		async function* foo() {
			yield 20;
		}
		expect(print(foo)).toMatchSnapshot();
		expect(print({ method: foo })).toMatchSnapshot();
	});

	it("prints properties", () => {
		function foo(a, b) {
			return a + b;
		}
		expect(print(foo, { all: true })).toMatchSnapshot();
		foo.bar = "Bar";
		expect(print(foo, { all: true })).toMatchSnapshot();
	});

	it("doesn't print source code if `noSource` is set", () => {
		function foo(i) {
			return i + 1;
		}
		function* bar() {
			yield true;
		}
		async function baz(i) {
			return i + 2;
		}
		async function* qux() {
			yield true;
		}
		const opts = { noSource: true };
		expect(print(foo, opts)).toMatchSnapshot();
		expect(print(bar, opts)).toMatchSnapshot();
		expect(print(baz, opts)).toMatchSnapshot();
		expect(print(qux, opts)).toMatchSnapshot();

		opts.all = true;
		expect(print(foo, opts)).toMatchSnapshot();
		expect(print(bar, opts)).toMatchSnapshot();
		expect(print(baz, opts)).toMatchSnapshot();
		expect(print(qux, opts)).toMatchSnapshot();
	});

	describe("Accessors", () => {
		it("prints getter/setter pairs", () => {
			let getCalls = 0;
			let setCalls = 0;
			expect(
				print({
					abc: "ABC",
					get foo() {
						return "Foo";
					},
					set foo(to) {
						setCalls++;
					},
					xyz: "XYZ",
				})
			).toMatchSnapshot();
			expect(setCalls).toEqual(0);

			const something = Symbol("something");
			getCalls = 0;
			setCalls = 0;
			expect(
				print({
					abc: "ABC",
					get [something]() {
						getCalls++;
						return true;
					},
					set [something](to) {
						setCalls++;
					},
					xyz: "XYZ",
				})
			).toMatchSnapshot();
			expect(getCalls).toEqual(0);
			expect(setCalls).toEqual(0);
		});

		it("doesn't print them when invoking getters", () => {
			let getterCalls = 0;
			let setterCalls = 0;
			expect(
				print(
					{
						abc: "ABC",
						get foo() {
							return ++getterCalls;
						},
						set foo(to) {
							setterCalls++;
						},
						xyz: "XYZ",
					},
					{ followGetters: true }
				)
			).toMatchSnapshot();
			expect(getterCalls).toEqual(1);
			expect(setterCalls).toEqual(0);
		});

		it("always prints write-only accessors", () => {
			const obj = {
				foo: "Foo",
				set name(to) {
					obj.foo = to;
				},
				bar: "Bar",
			};
			expect(print(obj, { followGetters: true })).toMatchSnapshot();
		});

		it("identifies references", () => {
			function bar(a, b) {
				return a + b;
			}
			const obj = Object.defineProperty({ a: 1, b: 2 }, "foo", {
				enumerable: true,
				get: bar,
				set: bar,
			});
			expect(print(obj)).toMatchSnapshot();
			bar.baz = "Qux";
			obj.c = 3;
			expect(print(obj)).toMatchSnapshot();
		});
	});

	describe("Line terminators", () => {
		const lines = [
			"function foo(c){",
			"\tlet a = 1;",
			"\tlet b = 2;",
			"\tif(a + b < c)",
			"\t\treturn a + b;",
			"}",
		];
		const fn = (eol) =>
			expect(print(Function("return " + lines.join(eol))())).toMatchSnapshot();
		it("prints source code that uses LF endings", () => {
			fn("\n");
		});
		it("prints source code that uses CR endings", () => {
			fn("\r");
		});
		it("prints source code that uses CRLF endings", () => {
			fn("\r\n");
		});
		it("prints source code that uses LS endings", () => {
			fn("\u2028");
		});
		it("prints source code that uses PS endings", () => {
			fn("\u2029");
		});
		it("prints source code with mixed endings", () => {
			const eol = ["\n", "\r", "\r\n", "\u2028", "\u2029"];
			for (let i = 0; i < 5; ++i) {
				const src = lines
					.map((line, index) => line + eol[(i + index) % 5])
					.join("");
				expect(print(Function("return " + src)())).toMatchSnapshot();
			}
		});
	});
});
