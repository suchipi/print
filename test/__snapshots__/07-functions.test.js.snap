// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Functions Accessors always prints write-only accessors 1`] = `
{
	foo: "Foo"
	set name: Function {
		│1│ set name(to) {
		│2│           obj.foo = to;
		│3│         }
	}
	bar: "Bar"
}
`;

exports[`Functions Accessors doesn't print them when invoking getters 1`] = `
{
	abc: "ABC"
	foo: 1
	xyz: "XYZ"
}
`;

exports[`Functions Accessors identifies references 1`] = `
{
	a: 1
	b: 2
	get foo: Function {
		│1│ function bar(a, b) {
		│2│         return a + b;
		│3│       }
	}
	set foo: -> {root}.get foo
}
`;

exports[`Functions Accessors identifies references 2`] = `
{
	a: 1
	b: 2
	get foo: Function {
		baz: "Qux"
		
		│1│ function bar(a, b) {
		│2│         return a + b;
		│3│       }
	}
	set foo: -> {root}.get foo
	c: 3
}
`;

exports[`Functions Accessors prints getter/setter pairs 1`] = `
{
	abc: "ABC"
	get foo: Function {
		│1│ get foo() {
		│2│           return "Foo";
		│3│         }
	}
	set foo: Function {
		│1│ set foo(to) {
		│2│           setCalls++;
		│3│         }
	}
	xyz: "XYZ"
}
`;

exports[`Functions Accessors prints getter/setter pairs 2`] = `
{
	abc: "ABC"
	xyz: "XYZ"
	get Symbol(something): Function {
		│1│ get [something]() {
		│2│           getCalls++;
		│3│           return true;
		│4│         }
	}
	set Symbol(something): Function {
		│1│ set [something](to) {
		│2│           setCalls++;
		│3│         }
	}
}
`;

exports[`Functions Line terminators prints source code that uses CR endings 1`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code that uses CRLF endings 1`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code that uses LF endings 1`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code that uses LS endings 1`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code that uses PS endings 1`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code with mixed endings 1`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code with mixed endings 2`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code with mixed endings 3`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code with mixed endings 4`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions Line terminators prints source code with mixed endings 5`] = `
Function {
	│1│ function foo(c){
	│2│ 	let a = 1;
	│3│ 	let b = 2;
	│4│ 	if(a + b < c)
	│5│ 		return a + b;
	│6│ }
}
`;

exports[`Functions doesn't print source code if \`noSource\` is set 1`] = `Function {}`;

exports[`Functions doesn't print source code if \`noSource\` is set 2`] = `GeneratorFunction {}`;

exports[`Functions doesn't print source code if \`noSource\` is set 3`] = `AsyncFunction {}`;

exports[`Functions doesn't print source code if \`noSource\` is set 4`] = `AsyncGeneratorFunction {}`;

exports[`Functions doesn't print source code if \`noSource\` is set 5`] = `
Function {
	length: 1
	name: "foo"
	arguments: null
	caller: null
	prototype: {
		constructor: -> {root}
	}
}
`;

exports[`Functions doesn't print source code if \`noSource\` is set 6`] = `
GeneratorFunction {
	length: 0
	name: "bar"
	prototype: Generator {}
}
`;

exports[`Functions doesn't print source code if \`noSource\` is set 7`] = `
AsyncFunction {
	length: 1
	name: "baz"
}
`;

exports[`Functions doesn't print source code if \`noSource\` is set 8`] = `
AsyncGeneratorFunction {
	length: 0
	name: "qux"
	prototype: AsyncGenerator {}
}
`;

exports[`Functions identifies asynchronous functions 1`] = `
AsyncFunction {
	│1│ async function foo() {
	│2│       return 20;
	│3│     }
}
`;

exports[`Functions identifies asynchronous functions 2`] = `
{
	method: AsyncFunction {
		│1│ async function foo() {
		│2│       return 20;
		│3│     }
	}
}
`;

exports[`Functions identifies asynchronous generators 1`] = `
AsyncGeneratorFunction {
	│1│ async function* foo() {
	│2│       yield 20;
	│3│     }
}
`;

exports[`Functions identifies asynchronous generators 2`] = `
{
	method: AsyncGeneratorFunction {
		│1│ async function* foo() {
		│2│       yield 20;
		│3│     }
	}
}
`;

exports[`Functions identifies generator functions 1`] = `
GeneratorFunction {
	│1│ function* foo() {
	│2│       yield 20;
	│3│     }
}
`;

exports[`Functions identifies generator functions 2`] = `
{
	method: GeneratorFunction {
		│1│ function* foo() {
		│2│       yield 20;
		│3│     }
	}
}
`;

exports[`Functions prints properties 1`] = `
Function {
	length: 2
	name: "foo"
	arguments: null
	caller: null
	prototype: {
		constructor: -> {root}
	}
	
	│1│ function foo(a, b) {
	│2│       return a + b;
	│3│     }
}
`;

exports[`Functions prints properties 2`] = `
Function {
	length: 2
	name: "foo"
	arguments: null
	caller: null
	prototype: {
		constructor: -> {root}
	}
	bar: "Bar"
	
	│1│ function foo(a, b) {
	│2│       return a + b;
	│3│     }
}
`;

exports[`Functions prints source code 1`] = `
Function {
	│1│ function func(i) {
	│2│       return i + 2;
	│3│     }
}
`;

exports[`Functions prints source code 2`] = `
Function {
	│1│ () => "A"
}
`;

exports[`Functions prints source code 3`] = `
{
	func: Function {
		│1│ function fn(value) {
		│2│         return value;
		│3│       }
	}
	arrow: Function {
		│1│ (a, b) => a + b
	}
}
`;
